### DISCUSSION TITLE
Search (Query) parameters

### DISCUSSION THREAD

**Discussion by richardrandak** [FortnoxAB/csharp-api-sdk#194] [General]
### Description
I did some tests/research around search parameters.
Here are some notes to think of

1) As stated in the [Fortnox documentation](https://developer.fortnox.se/general/parameters/), search parameters are used on server side in database/sql `LIKE` operation. Server also adds an `%` to both start and end of the parameter value.
Example: Specified parameter `123` ends up being `%123%` before used in database operation.

2) LIKE operation supports two wild card characters. These can be used in the search parameter value as well
`%` - representing an optional sub-string similar to regex `.*`
`_` - representing any character similar to regex `.`

3) The wildcard characters can be escaped to search for a value containing the character `_` or `%` by using slash, e.g. `\_` or `\%`

### Problems
It can be very confusing sometimes when a single entity is expected, but several are retrieved by the query.
Examples: 
1)
Parameter value: 123
Matches: 123, 1234, 1235 or even A123
2) 
Parameter value AB_12
Matches: AB_12, ABC12, AB112, or even XABC12X

### Conclusion
The problem I see is that there is no way to specify that user wants only exact matches. Something like `^` and `$` in regex. 
I believe that the server-side "feature" which wraps the parameter with `%` is a bad idea and it should be up to user to specify them if needed. This way the user needs to filter the results explicitly after retrieving them to get only the exact matches.

Also, the documentation does not mention that the `_` and `%` needs to be escaped. It is something I found out just now by digging into it.

What do you think?

### Open questions
1) What will happen if I set value to `ABC\` ? After server adds the percent characters e.g. `%ABC\%`, will the ending wildcard actually be escaped? :)

**Comment by bpaolo71** [comment]
I tried another experiment with Supplier Invoice File Connections
The request is `https://api.fortnox.se/3/supplierinvoicefileconnections/?supplierinvoicenumber=1`
according to what you said I would aspect to get a response with different results because the `%1%` on server side.
Instead I only got one result! This mens they don't always use the same criteria.

**Comment by richardrandak** [comment]
Interesting.. and confusing :) 
I was doing my experiments using Customer's Name property. When I find some more time, I will check other endpoints. Maybe there is different behavior if an "foreign key" is used in search (like in your case with supplier invoice number)

**Comment by enriheba** [comment]
Hi

We are creating a transfer of time to Fortnox salury.  
We have found the Saluryconnector, but we cant find the under table where time should be saved
Could you help us with tablename ?

**Reply by richardrandak** [reply]
@enriheba Check the `AttendanceTransactionsConnector` and `AbsenceTransactionsConnector`
